loom/engine/bindings/loom/lmSQLite.cpp

```diff
@@ -0,0 +1,647 @@
+    }
+    else
+    {
+        //no separator, so we need to prefix the system writable path in front of the database name
+        c->databaseFullPath = utString(cocos2d::CCFileUtils::sharedFileUtils()->getWriteablePath().c_str()) + c->databaseName;
+    }
+    
+    //open the SQLite DB
+    res = sqlite3_open_v2(c->databaseFullPath.c_str(), &c->dbHandle, flags, 0);
+    if(res != SQLITE_OK)
+    {
+        lmLogError(gSQLiteGroup, "Error opening the SQLite database file: %s with message: %s", c->databaseFullPath.c_str(), c->getErrorMessage());
+    }
+    return c;
+}
+
+bool Connection::backgroundImport(const char *database, JSON *data)
+{
+    if(Connection::backgroundImportInProgress)
+    {
+        lmLogError(gSQLiteGroup, "Attempting to run multiple backgroundImport calls on the same database: %s", database);
+        return false;
+    }
+
+    //store values to be used in the thread
+    Connection::backgroundImportDatabase = database;
+    Connection::backgroundImportData = data;
+
+    //start up the backgroundImport thread
+    Connection::backgroundImportInProgress = true;
+    loom_thread_start(Connection::backgroundImportBody, NULL);    
+    return true;
+}
+
+const char* Connection::getVersion()
+{
+    return sqlite3_libversion();
+}
+
+void Connection::backgroundImportDone(int result)
+{    
+    //fire our completion delegate
+    Connection::_OnImportCompleteDelegate.pushArgument(result);
+    Connection::_OnImportCompleteDelegate.invoke();
+
+    //we're done now, so need to turn off our processing flag
+    loom_mutex_lock(Connection::backgroundImportMutex);
+    Connection::backgroundImportInProgress = false;
+    loom_mutex_unlock(Connection::backgroundImportMutex);
+}
+
+int __stdcall Connection::backgroundImportBody(void *param)
+{
+    int i, j;
+    int numRows;
+    JSON *data;
+    JSON *table;
+    JSON *row;
+    Connection *c;
+    Statement *s;
+    utString query;
+    const char *database;
+    const char *tableName;
+    const char *columnName;
+    const char *firstColumn;
+
+
+    //get the database name and JSON data
+    database = Connection::backgroundImportDatabase;
+    data = Connection::backgroundImportData;
+
+///////////
+///////////
+//BEN QUESTION: Do we need to handle CREATE of a database as well? 
+//    Or will backgroundImport only be used on already created DBs?
+///////////
+///////////
+    
+    //open the database for read/write
+    c = Connection::open(database, SQLITE_OPEN_READWRITE);
+    if((c == NULL) || (c->getErrorCode() != SQLITE_OK))
+    {
+        if(c != NULL)
+        {
+            c->close();
+        }
+        Connection::backgroundImportDone(c->getErrorCode());        
+        return 0;
+    }
+    //NOTE: shouldn't need to yield the thread here as the thread will die now anyways...
+    // loom_thread_yield();
+
+    //pull out the JSON data to use in the query
+    //NOTE: this expects the JSON to be formatted as like:
```

Comment:
Yes, in the ls file for the calling function.
On 17 Feb 2015 3:31 AM, "Ben Garney" <notifications@github.com> wrote:

> In loom/engine/bindings/loom/lmSQLite.cpp
> <https://github.com/LoomSDK/LoomSDK/pull/261#discussion_r24786991>:
>
> > +    //open the database for read/write
> > +    c = Connection::open(database, SQLITE_OPEN_READWRITE);
> > +    if((c == NULL) || (c->getErrorCode() != SQLITE_OK))
> > +    {
> > +        if(c != NULL)
> > +        {
> > +            c->close();
> > +        }
> > +        Connection::backgroundImportDone(c->getErrorCode());
> > +        return 0;
> > +    }
> > +    //NOTE: shouldn't need to yield the thread here as the thread will die now anyways...
> > +    // loom_thread_yield();
> > +
> > +    //pull out the JSON data to use in the query
> > +    //NOTE: this expects the JSON to be formatted as like:
>
> Is this documented in the API docs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LoomSDK/LoomSDK/pull/261/files#r24786991>.
>